lib/db/backend/badger_backend.go

```diff
@@ -0,0 +1,377 @@
+	if err != nil {
+		panic(err)
+	}
+	return &badgerBackend{
+		bdb:    bdb,
+		closed: make(chan struct{}),
+	}
+}
+
+// badgerBackend implements Backend on top of a badger
+type badgerBackend struct {
+	bdb     *badger.DB
+	closeWG sync.WaitGroup
+	closed  chan struct{}
+}
+
+func (b *badgerBackend) NewReadTransaction() (ReadTransaction, error) {
+	return badgerSnapshot{
+		txn: b.bdb.NewTransaction(false),
+		rel: newReleaser(&b.closeWG),
+	}, nil
+}
+
+func (b *badgerBackend) NewWriteTransaction() (WriteTransaction, error) {
+	// We use two transactions here to preserve the property that our
+	// leveldb wrapper has, that writes in a transaction are completely
+	// invisible until it's committed, even inside that same transaction.
+	rtxn := b.bdb.NewTransaction(false)
+	wtxn := b.bdb.NewTransaction(true)
+	return &badgerTransaction{
+		badgerSnapshot: badgerSnapshot{
+			txn: rtxn,
+			rel: newReleaser(&b.closeWG),
+		},
+		txn: wtxn,
+		bdb: b.bdb,
+		rel: newReleaser(&b.closeWG),
+	}, nil
+}
+
+func (b *badgerBackend) Close() error {
+	b.closeWG.Wait()
+	err := b.bdb.Close()
+	close(b.closed)
+	return wrapBadgerErr(err)
+}
+
+func (b *badgerBackend) Get(key []byte) ([]byte, error) {
+	txn := b.bdb.NewTransaction(false)
+	defer txn.Discard()
+	item, err := txn.Get(key)
+	if err != nil {
+		return nil, wrapBadgerErr(err)
+	}
+	val, err := item.ValueCopy(nil)
+	if err != nil {
+		return nil, wrapBadgerErr(err)
+	}
+	return val, nil
+}
+
+func (b *badgerBackend) NewPrefixIterator(prefix []byte) (Iterator, error) {
+	select {
+	case <-b.closed:
+		return nil, errClosed{}
+	default:
+	}
+
+	txn := b.bdb.NewTransaction(false)
+	it := badgerPrefixIterator(txn, prefix)
+	it.releaseFn = func() {
+		txn.Discard()
+	}
+	return it, nil
+}
+
+func (b *badgerBackend) NewRangeIterator(first, last []byte) (Iterator, error) {
+	select {
+	case <-b.closed:
+		return nil, errClosed{}
+	default:
+	}
+
+	txn := b.bdb.NewTransaction(false)
+	it := badgerRangeIterator(txn, first, last)
+	it.releaseFn = func() {
+		txn.Discard()
+	}
+	return it, nil
+}
+
+func (b *badgerBackend) Put(key, val []byte) error {
+	select {
+	case <-b.closed:
+		return errClosed{}
+	default:
+	}
+
+	txn := b.bdb.NewTransaction(true)
+	if err := txn.Set(key, val); err != nil {
+		txn.Discard()
+		return wrapBadgerErr(err)
+	}
+	return wrapBadgerErr(txn.Commit())
+}
+
+func (b *badgerBackend) Delete(key []byte) error {
+	select {
+	case <-b.closed:
+		return errClosed{}
+	default:
+	}
+
+	txn := b.bdb.NewTransaction(true)
+	if err := txn.Delete(key); err != nil {
+		txn.Discard()
+		return wrapBadgerErr(err)
+	}
+	return wrapBadgerErr(txn.Commit())
+}
+
+func (b *badgerBackend) Compact() error {
+	select {
+	case <-b.closed:
+		return errClosed{}
+	default:
+	}
+
+	// XXX: check if this is appropriate or if we also need Flatten and
+	// whatnot. Also it apparently returns an error if the GC didn't result
+	// in anything being freed...
+	_ = b.bdb.RunValueLogGC(0.5)
```

Comment:
Is the return error typed in a way we can check it? Dropping this on the floor sucks.